contracts/registry/src/lib.rs

```diff
@@ -723,12 +742,175 @@ mod tests {
+        assert!(batch_metadata.len() == 100);
 
-        // tests:
-        // + test soult transfer with "continuation" - use the internal _sbt_soul_transfer to control limit
-        // + edge case: caller doesn't have any token
-        // + edge case: soul_transfer resumed but there is no more tokens (must emit event in the last resumed call)
-        // + find all edge cases
+        // issuer_1
+        ctr.sbt_mint(vec![(alice(), batch_metadata[..50].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(alice(), issuer1(), None), 50);
+
+        // issuer_2
+        ctx.predecessor_account_id = issuer2();
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(alice(), batch_metadata[50..].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(alice(), issuer2(), None), 50);
+
+        // add more tokens to issuer_1
+        ctx.predecessor_account_id = issuer1();
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(bob(), batch_metadata[..20].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(bob(), issuer1(), None), 20);
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(alice2(), batch_metadata[..20].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(alice2(), issuer1(), None), 20);
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(carol(), batch_metadata[..20].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(carol(), issuer1(), None), 20);
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(dan(), batch_metadata[..10].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(dan(), issuer1(), None), 10);
+
+        // soul transfer alice->alice2
+        ctx.predecessor_account_id = alice();
+
+        let limit: u32 = 25; //anything above this limit will fail due to exceeding maximum gas usage per call
+        loop {
+            ctx.prepaid_gas = max_gas;
+            testing_env!(ctx.clone());
+            let result = ctr._sbt_soul_transfer(alice2(), limit as usize);
+            if result.1 == true {
+                break;
+            }
+        }
+
+        // check all the balances afterwards
+        assert_eq!(ctr.sbt_supply_by_owner(alice(), issuer1(), None), 0);
+        assert_eq!(ctr.sbt_supply_by_owner(alice(), issuer2(), None), 0);
+        assert_eq!(ctr.sbt_supply_by_owner(alice2(), issuer1(), None), 70);
+        assert_eq!(ctr.sbt_supply_by_owner(alice2(), issuer2(), None), 50);
+        assert_eq!(ctr.sbt_supply_by_owner(bob(), issuer1(), None), 20);
+        assert_eq!(ctr.sbt_supply_by_owner(carol(), issuer1(), None), 20);
+        assert_eq!(ctr.sbt_supply_by_owner(dan(), issuer1(), None), 10);
+    }
+
+    #[test]
+    fn soul_transfer_limit_basics() {
+        let max_gas: Gas = Gas::ONE_TERA.mul(300);
```

Comment:
seams we are reusing this, let's declare it in the `mod test { ...`

Reply:
DONE